src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1080,38 +1080,60 @@ static SortDescription getSortDescriptionFromGroupBy(const ASTSelectQuery & quer
     return order_descr;
 }
 
-static UInt64 getLimitUIntValue(const ASTPtr & node, const ContextPtr & context, const std::string & expr)
+static bool getLimitUIntValue(const ASTPtr & node, const ContextPtr & context, const std::string & expr, UInt64 & res)
 {
     const auto & [field, type] = evaluateConstantExpression(node, context);
 
     if (!isNativeNumber(type))
         throw Exception(
             "Illegal type " + type->getName() + " of " + expr + " expression, must be numeric type", ErrorCodes::INVALID_LIMIT_EXPRESSION);
 
-    Field converted = convertFieldToType(field, DataTypeUInt64());
+    Field converted = convertFieldToType(field, DataTypeInt128());
```

Comment:
Why we cast it to `DataTypeInt128` ?

Reply:
As far as I understand, because of backward compatibility. We need to support previous range `[0, 2^64 - 1]` and support negative values, so we use `Int128` instead of `Int64`.

Reply:
Yes, I think we should treat limit and offset as UInt64 as before. And to support negative limit and offset, Int128 is used to get the sign of limit and offset value. So Int128 here.